File: kotlin/src/main/kotlin/problem18/4sum.kt
package problem18

fun main() {
    println(Solution().fourSum(intArrayOf(0, 2, 2, 2, 10, -3, -9, 2, -10, -4, -9, -2, 2, 8, 7), 6))
}

class Solution {
    fun fourSum(nums: IntArray, target: Int): List<List<Int>> {
        val l = nums.size
        val twoSums = mutableMapOf<Int, MutableList<Pair<Int, Int>>>()
        for (i in 0..(l - 2)) {
            for (j in (i + 1)..(l - 1)) {
                twoSums.append(nums[i] + nums[j], i to j)
            }
        }

        val res: MutableSet<FourSums> = mutableSetOf()
        for ((key, value) in twoSums) {
            twoSums[target - key]?.let {
                for ((a, b) in value) {
                    for ((c, d) in it) {
                        if (key == -4) {
                            println("$a, $b, $c, $d")
                            println("${nums[a]}, ${nums[b]}, ${nums[c]}, ${nums[d]}")
                        }
                        if (a != c && a != d && b != c && b != d) {
                            res.add(FourSums(listOf(nums[a], nums[b], nums[c], nums[d])))
                        }
                    }
                }
            }
        }
        return res.toList().map { it.elems }
    }
}

fun MutableMap<Int, MutableList<Pair<Int, Int>>>.append(key: Int, value: Pair<Int, Int>) {
    val list = this[key] ?: mutableListOf()
    list.add(value)
    this[key] = list
}

class FourSums(val elems: List<Int>) {
    override fun equals(other: Any?): Boolean {
        if (other is FourSums) {
            return this.elems.containsAll(other.elems) && other.elems.containsAll(this.elems)
        }
        return false
    }

    override fun hashCode(): Int {
        return this.elems.map { it.hashCode() }.sum()
    }
}

File: go/explore/july2021/day16/4sum.go
package main

import "sort"

func main() {
	for _, s := range fourSum([]int{1, 0, -1, 0, -2, 2}, 0) {
		for _, n := range s {
			print(n)
		}
		println()
	}
}

const RESET = 1000000001

func fourSum(nums []int, target int) [][]int {
	n := len(nums)
	l1 := RESET
	l2 := RESET
	l3 := RESET
	l4 := RESET

	sort.Ints(nums)
	res := [][]int{}

	for i := 0; i < n-3; i++ {
		if l1 == nums[i] {
			continue
		}
		l1 = nums[i]
		l2 = RESET
		l3 = RESET
		l4 = RESET
		for j := i + 1; j < n-2; j++ {
			if l2 == nums[j] {
				continue
			}
			l2 = nums[j]
			l3 = RESET
			l4 = RESET

			l := n - 1
			for k := j + 1; k < l; k++ {
				if l3 == nums[k] {
					continue
				}
				l3 = nums[k]
				want := target - nums[i] - nums[j] - nums[k]
				for k < l && (nums[l] > want || nums[l] == l4) {
					l4 = nums[l]
					l--
				}
				if k == l {
					break
				}
				if want == nums[l] {
					res = append(res, []int{nums[i], nums[j], nums[k], nums[l]})
				}
			}
		}
	}
	return res
}
